File: Python/OneLang/StdLib/PackagesFolderSource.py
from OneLangStdLib import *
from OneFile import *
import OneLang.StdLib.PackageManager as packMan
import re

class PackagesFolderSource:
    def __init__(self, packages_dir = "packages"):
        self.packages_dir = packages_dir
    
    def get_package_bundle(self, ids, cached_only):
        raise Error("Method not implemented.")
    
    def get_all_cached(self):
        packages = {}
        all_files = OneFile.list_files(self.packages_dir, True)
        for fn in all_files:
            if fn == "bundle.json":
                continue
            # TODO: hack
            path_parts = re.split("/", fn)
            # [0]=implementations/interfaces, [1]=package-name, [2:]=path
            type = path_parts.pop(0)
            pkg_dir = path_parts.pop(0)
            if type != "implementations" and type != "interfaces":
                continue
            # skip e.g. bundle.json
            pkg_id_str = f'''{type}/{pkg_dir}'''
            pkg = packages.get(pkg_id_str)
            if pkg == None:
                pkg_dir_parts = re.split("-", pkg_dir)
                version = re.sub("^v", "", pkg_dir_parts.pop())
                pkg_type = packMan.PACKAGE_TYPE.IMPLEMENTATION if type == "implementations" else packMan.PACKAGE_TYPE.INTERFACE
                pkg_id = packMan.PackageId(pkg_type, "-".join(pkg_dir_parts), version)
                pkg = packMan.PackageContent(pkg_id, {}, True)
                packages[pkg_id_str] = pkg
            pkg.files["/".join(path_parts)] = OneFile.read_text(f'''{self.packages_dir}/{fn}''')
        return packMan.PackageBundle(packages.values())

File: PHP/src/OneLang/StdLib/PackagesFolderSource.php
<?php

namespace OneLang\StdLib\PackagesFolderSource;

use OneFile\OneFile;
use OneLang\StdLib\PackageManager\PackageSource;
use OneLang\StdLib\PackageManager\PackageId;
use OneLang\StdLib\PackageManager\PackageBundle;
use OneLang\StdLib\PackageManager\PackageContent;
use OneLang\StdLib\PackageManager\PackageType;

class PackagesFolderSource implements PackageSource {
    public $packagesDir;
    
    function __construct($packagesDir = "packages") {
        $this->packagesDir = $packagesDir;
    }
    
    function getPackageBundle($ids, $cachedOnly) {
        throw new \OneCore\Error("Method not implemented.");
    }
    
    function getAllCached() {
        $packages = Array();
        $allFiles = OneFile::listFiles($this->packagesDir, true);
        foreach ($allFiles as $fn) {
            if ($fn === "bundle.json")
                continue;
            // TODO: hack
            $pathParts = preg_split("/\\//", $fn);
            // [0]=implementations/interfaces, [1]=package-name, [2:]=path
            $type = array_shift($pathParts);
            $pkgDir = array_shift($pathParts);
            if ($type !== "implementations" && $type !== "interfaces")
                continue;
            // skip e.g. bundle.json
            $pkgIdStr = $type . "/" . $pkgDir;
            $pkg = @$packages[$pkgIdStr] ?? null;
            if ($pkg === null) {
                $pkgDirParts = preg_split("/-/", $pkgDir);
                $version = preg_replace("/^v/", "", array_pop($pkgDirParts));
                $pkgType = $type === "implementations" ? PackageType::IMPLEMENTATION : PackageType::INTERFACE_;
                $pkgId = new PackageId($pkgType, implode("-", $pkgDirParts), $version);
                $pkg = new PackageContent($pkgId, Array(), true);
                $packages[$pkgIdStr] = $pkg;
            }
            $pkg->files[implode("/", $pathParts)] = OneFile::readText($this->packagesDir . "/" . $fn);
        }
        return new PackageBundle(array_values($packages));
    }
}
